data/blog/js-mutation-testing-our-sdks.md

```diff
@@ -0,0 +1,166 @@
+So playing devil's advocate, you could say coverage really only tells you one thing with certainty: Which parts of your code are _not_ covered by any test.
+
+Let's circle back to the part about coverage not telling you if an introduced bug would be caught. What if there was a tool that does exactly that? Enter mutation testing.
+
+The idea of mutation testing is fairly simple:
+
+1. Make a slight modification to your codebase. (In MT terms: Apply a _mutant operator_ to your code base.) For example, change an equality operator:
+   ```js
+   function isEven(num) {
+     // return num % 2 === 0; // mutate to:
+     return num % 2 !== 0
+   }
+   ```
+   The resulting code is called a "mutant".
+2. Run your tests against the mutant.
+   - If at least one of your test fails, crashes or times out, the mutant was _killed_.
+   - If all your tests still pass without a crash or timeout, the mutant _survived_.
+3. Repeat 1 and 2 until all possible mutant operators have been applied to your codebase or you reached a pre-defined maximum number of mutants.
+
+After this virtual bloodbath, you can calculate a _Mutation Score_ by dividing the number of killed mutants by the number of all created mutants. The score now tells you how likely it is that your tests would catch an actual bug.
+
+### Mutant Similarity Assumption
+
+The entire concept of MT depends on the assumption that [mutants are similar to actual bugs](https://dl.acm.org/doi/abs/10.1145/2635868.2635929) that programmers might introduce into the code base.
+Now the question is, is this true in reality? Luckily, engineers at Google publish papers about mutation testing every now and then. In [one of them](https://dl.acm.org/doi/10.1109/ICSE43902.2021.00087), they
+investigated the similarity assumption by comparing 15 million mutants against bugfix PRs. They found a high similarity of mutants to bugs, suggesting that the assumption does hold up.
+Does this translate to other code bases? Maybe not, but for the sake of trying things out, let's go with "yes".
+
+### Performance
+
+You might have already guessed it but what really keeps mutation testing from being widely adopted is its high performance impact. Running 1000s of tests on 1000s of mutants is computationally expensive - much higher than any kind of coverage calculation. Fortunately, modern MT tools provide optimizations that minimize the runtime as much as possible. For example, they use "per-test" mutation coverage to determine which tests cover a mutant and then, only execute this subset of all tests on the specific mutant. Similarly, modern MT tools chain multiple mutants into one modified code base and activate a specific mutant at runtime. This means that the modified code base doesn't have to be re-built for every mutant, but instead only once at the beginning of the MT run.
```

Comment:
```suggestion
What really keeps mutation testing from being widely adopted is its high performance impact. Running 1000s of tests on 1000s of mutants is computationally expensive - much higher than any kind of coverage calculation. Fortunately, modern MT tools provide optimizations that minimize the runtime as much as possible. For example, they use "per-test" mutation coverage to determine which tests cover a mutant and then, only execute this subset of all tests on the specific mutant. Similarly, modern MT tools chain multiple mutants into one modified code base and activate a specific mutant at runtime. This means that the modified code base doesn't have to be re-built for every mutant, but instead only once at the beginning of the MT run.
```

not sure if people might have guessed this - best to avoid assumptions in writing ;)

Reply:
Thanks for the feedback, agree with both items! I opened https://github.com/getsentry/sentry.engineering/pull/142 with the rewordings.